### PR TITLE
[Only review]Fix #733 the pipeline parameter consumed as file

### LINKED ISSUE #733: Data passing doesn't work if the pipeline parameter that get consumed as file.
/kind bug

**Issue**
1. Here is a simple KFP application:
```
import kfp
from kfp.components import create_component_from_func, InputPath
from kfp_tekton.compiler import TektonCompiler

# Consume as file
@create_component_from_func
def consume_anything_as_file(data_path: InputPath()):
    with open(data_path) as f:
        print("consume_anything_as_file: " + f.read())

# Pipeline
@kfp.dsl.pipeline(name='data_passing_pipeline')
def data_passing_pipeline(
    anything_param="anything_param",
):
    # Pass pipeline parameter; consume as file
    consume_anything_as_file(anything_param)

if __name__ == "__main__":
    kfp_endpoint = None
    TektonCompiler().compile(data_passing_pipeline, __file__.replace('.py', '.yaml'))
```
2. The generated yaml file is below:
```
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: data-passing-pipeline
  annotations:
    tekton.dev/output_artifacts: '{}'
    tekton.dev/input_artifacts: '{}'
    tekton.dev/artifact_bucket: mlpipeline
    tekton.dev/artifact_endpoint: minio-service.kubeflow:9000
    tekton.dev/artifact_endpoint_scheme: http://
    tekton.dev/artifact_items: '{"consume-anything-as-file": []}'
    sidecar.istio.io/inject: "false"
    pipelines.kubeflow.org/pipeline_spec: '{"inputs": [{"default": "anything_param",
      "name": "anything_param", "optional": true}], "name": "data_passing_pipeline"}'
spec:
  params:
  - name: anything_param
    value: anything_param
  pipelineSpec:
    params:
    - name: anything_param
      default: anything_param
    tasks:
    - name: consume-anything-as-file
      taskSpec:
        steps:
        - name: main
          args:
          - --data
          - $(workspaces.consume-anything-as-file.path)/anything_param
          command:
          - sh
          - -ec
          - |
            program_path=$(mktemp)
            printf "%s" "$0" > "$program_path"
            python3 -u "$program_path" "$@"
          - |
            def consume_anything_as_file(data_path):
                with open(data_path) as f:
                    print("consume_anything_as_file: " + f.read())

            import argparse
            _parser = argparse.ArgumentParser(prog='Consume anything as file', description='')
            _parser.add_argument("--data", dest="data_path", type=str, required=True, default=argparse.SUPPRESS)
            _parsed_args = vars(_parser.parse_args())

            _outputs = consume_anything_as_file(**_parsed_args)
          image: python
          imagePullPolicy: IfNotPresent
        metadata:
          labels:
            pipelines.kubeflow.org/pipelinename: ''
            pipelines.kubeflow.org/generation: ''
            pipelines.kubeflow.org/cache_enabled: "true"
          annotations:
            pipelines.kubeflow.org/component_spec: '{"implementation": {"container":
              {"args": ["--data", {"inputPath": "data"}], "command": ["sh", "-ec",
              "program_path=$(mktemp)\nprintf \"%s\" \"$0\" > \"$program_path\"\npython3
              -u \"$program_path\" \"$@\"\n", "def consume_anything_as_file(data_path):\n    with
              open(data_path) as f:\n        print(\"consume_anything_as_file: \"
              + f.read())\n\nimport argparse\n_parser = argparse.ArgumentParser(prog=''Consume
              anything as file'', description='''')\n_parser.add_argument(\"--data\",
              dest=\"data_path\", type=str, required=True, default=argparse.SUPPRESS)\n_parsed_args
              = vars(_parser.parse_args())\n\n_outputs = consume_anything_as_file(**_parsed_args)\n"],
              "image": "python:3.7"}}, "inputs": [{"name": "data"}], "name": "Consume
              anything as file"}'
            tekton.dev/template: ''
        workspaces:
        - name: consume-anything-as-file
      timeout: 0s
      workspaces:
      - name: consume-anything-as-file
        workspace: data-passing-pipeline
    workspaces:
    - name: data-passing-pipeline
  timeout: 0s
  workspaces:
  - name: data-passing-pipeline
    volumeClaimTemplate:
      spec:
        accessModes:
        - ReadWriteMany
        resources:
          requests:
            storage: 2Gi
        storageClassName: ""
```
3. The pvc and pv is ready:
```
[root@host xx]# kubectl get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                     STORAGECLASS   REASON   AGE
pv-data-passing                            20Gi       RWX            Recycle          Bound    default/pvc-5af0e90116                            46m
pvc-68ca9226-240e-42d3-8612-5e7d264f5aa1   20Gi       RWO            Delete           Bound    kubeflow/minio-pvc        standard                22d
pvc-d50c8532-ceaa-4e29-bbf4-fb844b5ddc97   20Gi       RWO            Delete           Bound    kubeflow/mysql-pv-claim   standard                22d
[root@host xx]# kubectl get pvc
NAME             STATUS   VOLUME            CAPACITY   ACCESS MODES   STORAGECLASS   AGE
pvc-5af0e90116   Bound    pv-data-passing   20Gi       RWX                           43m
[root@host xx]# 
```
4. When run the yaml, but the result is:
```
[root@dstest11 liuhui]# kubectl get po
NAME                                                             READY   STATUS   RESTARTS   AGE
data-passing-pipeline-consume-anything-as-file-4q8lh-pod-jlcc5   0/1     Error    0          44m
[root@dstest11 liuhui]# kubectl logs data-passing-pipeline-consume-anything-as-file-4q8lh-pod-jlcc5
Traceback (most recent call last):
  File "/tmp/tmp.dHojU3JdGf", line 10, in <module>
    _outputs = consume_anything_as_file(**_parsed_args)
  File "/tmp/tmp.dHojU3JdGf", line 2, in consume_anything_as_file
    with open(data_path) as f:
FileNotFoundError: [Errno 2] No such file or directory: '/workspace/consume-anything-as-file/anything_param'
[root@dstest11 liuhui]# 
```

**Expect Result:**
Should handle this case when use the pipeline parameter as file.

**Environment:**
* Python Version (use `python --version`):  Python 3.8.6
* SDK Version:  kfp-tekton SDK 1.0.0
* Tekton Version (use `tkn version`):
* Kubernetes Version (use `kubectl version`): v1.21.1
* OS (e.g. from `/etc/os-release`): RHEL8.4

### PR DESCRIPTION
**Pull request by huixa**
**Which issue is resolved by this Pull Request:** 
Resolves #733 

**Description of your changes:**
Remove the limitations for the pipeline parameter consumed as file.

**Environment tested:**

* Python Version (use `python --version`): Python 3.8.6
* Tekton Version (use `tkn version`):
* Kubernetes Version (use `kubectl version`): v1.21.1
* OS (e.g. from `/etc/os-release`): RHEL8.4

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by huixa** [comment]
@pugangxa Create PR and try to resolve the #733. And run the case as attachment, it could add step to use the pipeline parameter content to file, then be consumed as file. But because this file `_data_passing_rewriter.py` is related with other data passing's cases, maybe there has some impact. So just create this pr and please help guide me. Thanks.
[1.py.txt](https://github.com/kubeflow/kfp-tekton/files/7233758/1.py.txt)
[1.yaml.txt](https://github.com/kubeflow/kfp-tekton/files/7233759/1.yaml.txt)

**Comment by pugangxa** [comment]
Took a look and I think the root cause is that the parameter is in fact from pipeline parameter but not the dependent tasks, so what about match pipeline parameter and add the copy steps, like this:

```
diff --git a/sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py b/sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py
index 5bd8399..22ccbaa 100644
--- a/sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py
+++ b/sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py
@@ -17,6 +17,7 @@ import json
 import re

 from typing import List, Optional, Set

 from kfp_tekton.compiler._k8s_helper import sanitize_k8s_name
 from kfp_tekton.compiler._op_to_template import _get_base_step, _add_mount_path, _prepend_steps
@@ -300,7 +301,7 @@ def fix_big_data_passing(workflow: dict) -> dict:

     # Use workspaces to tasks if big data passing instead of 'results', 'copy-inputs'
     for task_template in container_templates:
-        task_template = big_data_passing_tasks(task_template,
+        task_template = big_data_passing_tasks(pipelinerun_name, task_template,
                                                pipelinerun_template,
                                                inputs_consumed_as_artifacts,
                                                outputs_consumed_as_artifacts)
@@ -470,7 +471,7 @@ def big_data_passing_pipelinerun(name: str, pr: dict, pw: set):
     return pr, prw


-def big_data_passing_tasks(task: dict, pipelinerun_template: dict,
+def big_data_passing_tasks(prname: str, task: dict, pipelinerun_template: dict,
                            inputs_tasks: set, outputs_tasks: set) -> dict:
     task_name = task.get('name')
     task_spec = task.get('taskSpec', {})
@@ -523,6 +524,10 @@ def big_data_passing_tasks(task: dict, pipelinerun_template: dict,
         if (task_name, task_artifact.get('name')) not in inputs_tasks:
             # add input artifact processes
             task = input_artifacts_tasks(task, task_artifact)
+        if (prname, task_artifact.get('name')) in inputs_tasks:
+            # add input artifact processes for pipeline parameter
+            task_artifact['raw']['data'] = pipelinerun_template['spec']['params'][0]['value']
+            task = input_artifacts_tasks(task, task_artifact)

     # Remove artifacts parameter from params
     task.get("taskSpec", {})['params'] = [
```
I tested it can add copy steps but still some issues in the code:
- Match the parameter values and add to the copy task
- Maybe do not change the function signature but get prname from pipelinerun_template
- Please test if there's multiple pipeline parameters but used as the same name in task_artifact, will duplicated items being created

Thanks

**Comment by pugangxa** [comment]
Thanks very much @huixa for your hard work. 
- Please sign the CLA
- Please add a test for this, can simply reuse the you python file with some small changes
- Please fix the lint error

File: sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py [hunk 1]
```diff
@@ -537,6 +548,24 @@ def big_data_passing_tasks(task: dict, pipelinerun_template: dict,
     return task
 
 
+def input_artifacts_tasks1(template: dict, artifact: dict) -> dict:
```
[the remark this answers is not in the document]
Comment (huixa) [reply sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py:551]: Thanks Gang. Yes, it's different function behavior for input_artifacts_tasks and input_artifacts_tasks1. I will update the input_artifacts_tasks1 name.

**Comment by huixa** [comment]
> Thanks very much @huixa for your hard work.
> 
>     * Please sign the CLA
> 
>     * Please add a test for this, can simply reuse the you python file with some small changes
> 
>     * Please fix the lint error

Thanks Gang very much. Will follow these issues.

**Comment by huixa** [comment]
@googlebot I signed it!

**Comment by huixa** [comment]
@Tomcli Had followed all the comments, could you help review again. Thanks.

**Comment by Tomcli** [comment]
thanks @huixa, can you run `make unit_test GENERATE_GOLDEN_YAML=True` to generate the test yaml because it will include the license and format it correctly.

For the python lint, you can run `make lint` to make sure no error is in the python test files.

**Comment by huixa** [comment]
> thanks @huixa, can you run `make unit_test GENERATE_GOLDEN_YAML=True` to generate the test yaml because it will include the license and format it correctly.
> 
> For the python lint, you can run `make lint` to make sure no error is in the python test files.

Thanks Tommy. Had fixed these issues and run these 2 commands successfully.
```
HuiMac:kfp-tekton liuhui$ make unit_test GENERATE_GOLDEN_YAML=True
==================================================================
Optional environment variables to configure unit_test, examples:
  make unit_test GENERATE_GOLDEN_YAML=True
============================================
... ...
test_data_passing_pipeline_param_as_file (compiler.compiler_tests.TestTektonCompiler)
Test compiling a pipeline_param_as_file workflow. ... ok
... ...

----------------------------------------------------------------------
Ran 73 tests in 9.344s

OK
unit_test: OK
HuiMac:kfp-tekton liuhui$ make lint
lint: OK
HuiMac:kfp-tekton liuhui$ 
```

**Comment by Tomcli** [comment]
thanks @huixa for your contributions.
/lgtm
/approve

**Comment by huixa** [comment]
@pugangxa Thank you for helping me analyze the root cause.
@Tomcli  Thank you for sharing the commands to resolve the unit test cases and python issues.
Finally appreciate Gang and Tommy again for this pr‘s review, guidance and help one step by step.

**Comment by pugangxa** [comment]
Congratulations! This is your first PR and is a good start point!

**Pull request merged**